source/slang/slang-emit.cpp

```diff
@@ -714,17 +714,35 @@ SlangResult emitEntryPointSourceFromIR(
         Session* session = compileRequest->getSession();
 
         // Get the downstream compiler needed for final target
-        PassThroughMode passThru = getDownstreamCompilerRequiredForTarget(targetRequest->target);
+        PassThroughMode passThru = getDownstreamCompilerRequiredForTarget(session, targetRequest->target);
 
-        // If generic CPP work out what compiler will actually be used
-        if (passThru == PassThroughMode::GenericCCpp)
+        // If nothing was *needed*, we still need some idea of which downstream compiler is going to be used for the
+        // prelude (the prelude is associated with each specific downstream compiler)
+        if (passThru == PassThroughMode::None)
         {
-            const SourceLanguage sourceLanguage = (sourceStyle == SourceStyle::C) ? SourceLanguage::C : SourceLanguage::CPP;
-            // Get the compiler used for the language
-            DownstreamCompiler* compiler = session->getDefaultDownstreamCompiler(sourceLanguage);
-            if (compiler)
+            // TODO(JS):
+            // NOTE! This makes the *assumption* that if we are outputting source, we don't want the
+            // prelude for a specific compiler, we are happy to just use the GenericCCpp
+            // If we didn't do this, we would have to do the work out what's available etc.
+            //
+            // This is all a bit unfortunate. The decision to associate preludes on compilers seemed like a good idea, but was perhaps a mistake.
+            // That it would be simpler if prelude was based on output source type.
```

Comment:
Yes, I think a follow-up change to make the prelude simply be determined by the `CodeGenTarget` could be nice to have.

Reply:
You would prefer CodeGenTarget over SlangSourceLanguage?

Reply:
Of the top of my head, `CodeGenTarget` seems more strictly correct, in that the prelude is always something we combine with generated (target) code.

I suppose one concern you might be thinking about is when the `CodeGenTarget` is the final destination target (e.g., `HostCallable`) while the prelude should be associated with the target we selected as a stepping stone to get there (e.g., `Cpp`). No matter what it seems like that scenario requires us to compute what the stepping-stone language/target should be, and the choice of whether that is a `CodeGenTarget` or a `SourceLanguage` isn't the most important.

(In practice, every textual language we support as a target is in *both* `CodeGenTarget` and `SourceLanguage`, so this is more of a philosophical point than anything else; either shoudl work)

Realistically, if we could revisit the fundamental design choices, it might be nice to have both `CodeGenTarget` and `SlangSourceLanguage` collapsed into a single `enum` that represents formats in which code can be expressed. Some day we might like Slang to be able to ingest something like a SPIR-V module as an input, at which point we'd either have to represent inputs using `CodeGenTarget` or extend `SlangSourceLanguage` to include things that aren't really "source" languages...